src/implementations/junoswap/staking.rs

```diff
@@ -0,0 +1,172 @@
+                    id: 0,
+                    msg: CosmosMsg::Wasm(WasmMsg::Execute {
+                        contract_addr: addr.to_string(),
+                        funds: vec![],
+                        msg: to_binary(&StakeCw20ExternalRewardsExecuteMsg::Claim {})?,
+                    }),
+                    gas_limit: None,
+                    reply_on: ReplyOn::Never,
+                })
+            })
+            .collect::<StdResult<Vec<SubMsg>>>()?;
+
+        let event = Event::new("apollo/cw-dex/claim_rewards").add_attribute("type", "junoswap");
+
+        Ok(Response::new().add_submessages(claim_messages).add_event(event))
+    }
+}
+
+impl Unlock for JunoswapStaking {
+    fn withdraw_unlocked(
+        &self,
+        _deps: Deps,
+        _env: &Env,
+        amount: Uint128,
+    ) -> Result<Response, CwDexError> {
+        let claim_unlocked_msg = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: self.addr.to_string(),
+            funds: vec![],
+            msg: to_binary(&Cw20StakeExecuteMsg::Claim {})?,
+        });
+
+        let event = Event::new("cw-dex/lockup/withdraw_unlocked")
+            .add_attribute("type", "junoswap")
+            .add_attribute("amount", amount.to_string());
+
+        Ok(Response::new().add_message(claim_unlocked_msg).add_event(event))
+    }
+
+    fn unlock(&self, _deps: Deps, _env: &Env, amount: Uint128) -> Result<Response, CwDexError> {
+        let unstake_msg = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: self.addr.to_string(),
+            funds: vec![],
+            msg: to_binary(&Cw20StakeExecuteMsg::Unstake {
+                amount,
+            })?,
+        });
+
+        let event = Event::new("cw-dex/staking/unstake")
+            .add_attribute("type", "junoswap")
+            .add_attribute("amount", amount.to_string());
+
+        Ok(Response::new().add_message(unstake_msg).add_event(event))
+    }
+}
+
+impl ForceUnlock for JunoswapStaking {
+    fn force_unlock(
+        &self,
+        _deps: Deps,
+        _env: &Env,
+        _lockup_id: Option<u64>,
+        _amount: Uint128,
+    ) -> Result<Response, CwDexError> {
+        Err(CwDexError::Std(StdError::generic_err("Cannot force unlock on Junoswap staking")))
+    }
+}
+
+impl LockedStaking for JunoswapStaking {
+    fn get_lockup_duration(&self, deps: Deps) -> Result<Duration, CwDexError> {
+        let duration = deps
+            .querier
+            .query::<Config>(&QueryRequest::Wasm(WasmQuery::Smart {
+                contract_addr: self.addr.to_string(),
+                msg: to_binary(&Cw20StakeQueryMsg::GetConfig {})?,
+            }))?
+            .unstaking_duration;
+
+        Ok(match duration {
+            Some(duration) => match duration {
+                cw_utils_0_13_4::Duration::Height(height) => Duration::Height(height),
+                cw_utils_0_13_4::Duration::Time(time) => Duration::Time(time),
+            },
+            None => Duration::Time(0),
```

Comment:
Does this assumption work with the logic in our vault contracts? Might be dangerous as the implementors of vaults don't realize that this can happen. Perhaps we should change the trait to return an `Option<Duration>`? That would also be a bit confusing since you would assume that any staking struct that implements the LockedStaking trait would have a duration though... Not sure here whats best.

Reply:
Right now we call `get_lockup_duration` in the the `QueryMsg::LockupDuration` handler and in the reply entrypoint for the OsmosisVault. Since the unlocking logic will be unique for Junoswap, I think this is okay as we can handle it explicitly there.